src/libs/merge/kdbmerge.c

```diff
@@ -609,6 +644,215 @@ static int checkSingleSet (KeySet * checkedSet, KeySet * firstCompared, KeySet *
+			keyDel (iterator);
+			return NULL;
+		}
+		strncat (result, keyString (lookup), size); // size of whole buffer
+		strcat (result, "\n");
+		if (counter >= 2)
+		{
+			int retval = keyDel (lookup);
+			if (retval != 0)
+			{
+				char msg[300];
+				if (retval < 0)
+				{
+					snprintf (msg, 300, "Could not delete key from key set because null pointer.");
+				}
+				else
+				{
+					snprintf (msg, 300, "Could not delete key with name %s from key set. There are %d references left.",
+						  keyName (lookup), retval);
+				}
+				ELEKTRA_SET_INTERNAL_ERROR (informationKey, msg);
+				elektraFree (result);
+				keyDel (iterator);
+				return NULL;
+			}
+		}
+		if (elektraArrayIncName (iterator) < 0)
+		{
+			ELEKTRA_SET_INTERNAL_ERROR (informationKey, "Increasing array key failed.");
+			elektraFree (result);
+			keyDel (iterator);
+			return NULL;
+		}
+	}
+	keyDel (iterator);
+	return result;
+}
+
+/**
+ * Turns an array from getValuesAsArray into a key set again.
+ * Frees the array and creates a new key set.
+ *
+ * @array array to be converted, has no \0 terminator at end
+ * @length length of array, does not include size for \0 as this is not there
+ * @informationKey for errors
+ * @returns the KeySet
+ */
+static KeySet * ksFromArray (const char * array, int length, Key * informationKey)
+{
+	if (array == NULL)
+	{
+		ELEKTRA_SET_INTERNAL_ERROR (informationKey, "Parameter must not be null.");
+		return NULL;
+	}
+	KeySet * result = ksNew (0, KS_END);
+	if (result == NULL)
+	{
+		ELEKTRA_SET_OUT_OF_MEMORY_ERROR (informationKey, "Memory allocation failed.");
+		return NULL;
+	}
+	Key * iterator = keyNew ("/#0", KEY_END); // TODO here
+	if (iterator == NULL)
+	{
+		ksDel (result);
+		ELEKTRA_SET_OUT_OF_MEMORY_ERROR (informationKey, "Memory allocation failed.");
+		return NULL;
+	}
+	char * buffer = elektraCalloc (length + 1); // + 1 for terminating null character
+	memcpy (buffer, array, length);
+	char * saveptr;
+	char * token = strtok_r (buffer, "\n", &saveptr);
+	do
+	{
+		ksAppendKey (result, keyNew (keyName (iterator), KEY_VALUE, token, KEY_END));
+		if (elektraArrayIncName (iterator) < 0)
+		{
+			ELEKTRA_SET_INTERNAL_ERROR (informationKey, "Increasing array key failed.");
+			// TODO more free?
+			elektraFree (result);
+			keyDel (iterator);
+			return NULL;
+		}
+	} while ((token = strtok_r (NULL, "\n", &saveptr)) != NULL);
+	elektraFree (buffer);
+	keyDel (iterator);
+	return result;
+}
+
+
+/**
+ * Removes all the arrays from our, their, base and result and puts the result of the merge into resultSet
+ * @param ourSet our
+ * @param theirSet their
+ * @param baseSet base
+ * @param resultSet result
+ * @retval 0 on success, -1 on error
+ */
+static int handleArrays (KeySet * ourSet, KeySet * theirSet, KeySet * baseSet, KeySet * resultSet, Key * informationKey)
+{
+	Key * checkedKey;
+	KeySet * toAppend = NULL;
+	while ((checkedKey = ksNext (baseSet)) != NULL)
+	{
+		if (elektraArrayValidateName (checkedKey) >= 0)
```

Comment:
I think it would make sense to simplify this code. Instead of:

```c
if (elektraArrayValidateName (checkedKey) >= 0) {
  // do something
}
```

you can use:

```c
if (elektraArrayValidateName (checkedKey) < 0) continue;
// do something
```

, as far as I can tell.

Reply:
I do not see where this change simplifies the code. It is rather a matter of style to use continue vs. deeper nesting. To put "do something" into a function (or several), however, might make the code simpler.

Reply:
It definitely simplifies the code from a human perspective in my opinion. This:

```c
if (!bla) return;
if (!blubb) return;

// do something
```

is easier to read, than

```c

if (bla) {
  if (blubb) {
  // do something
}
```

, especially if “do something” spans multiple lines.

Reply:
I understand that less nesting is somehow easier to read. However, there is still a lesson on structured progrmaming burned into my mind that said "One exit point for a function". This second option looks simple to me as well. I'm still torn between the two.

For this
```
if (bla) {
  if (blubb) {
  // do something
}
```
example we could also add a third variant
```
if (bla && blubb) {
  // do something
}
```
 which reduces indentation.


> To put "do something" into a function (or several), however, might make the code simpler.

Less indentation would make it more readable I guess, but I don't really see where I could split up the code in the body of `if (elektraArrayValidateName` in a simpler way.

Reply:
Yes, if you are not used to the style with multiple exit points it is dangerous. So better to not introduce it if the rest of your code does not do it like this.